disallow impl without base

dwblaikie opened the pull request:
- **Reject/error on base declarations that appear after field declarations**
- **Move the diagnostic handling a bit further up, near other diagnostics.**
- **Rephrase diagnostic**
- **`must come before` -> `must appear before`**
- **Add TODO for adding a note with an example field decl**
- **Error if a function is marked `impl` but there is no base class.**

dwblaikie committed:
Error if a function is marked `impl` but there is no base class.

dwblaikie commented:
Oops, this has some other stuff from another branch. Give me a sec to clean that up before reviewing.

dwblaikie commented:
Fixed.

zygoloid commented on toolchain/check/testdata/class/virtual_modifiers.carbon:
```diff
@@ -15,6 +15,10 @@ package Modifiers;
+  // CHECK:STDERR:   impl fn I();
+  // CHECK:STDERR:   ^~~~~~~~~~~~
+  // CHECK:STDERR:
   impl fn I();
```
Can you move the `impl fn` portions of this test out to a separate `// --- fail_modifiers.carbon` split-file within this test file?

(Background: we put passing and failing tests into different split-files within a test file. Split-files with errors should be named `fail_...` and ones without errors should not -- the test infrastructure checks this. The idea here is that if autoupdate causes a passing split-file to start failing or vice versa, we will catch that despite the autoupdate changing the expectations.)

zygoloid replied:
The tests seem to have got wedged somehow, but I'd expect they would fail for this reason.

dwblaikie replied:
Ah, yeah - updated this test as requested and mucked about with the other test cases since there was some loss of coverage of the non-fail cases now that the requirements were a bit more strict, so pulled those out into more split-file chunks in this file. Let's see if the testing unwedges...

zygoloid commented on toolchain/check/handle_function.cpp:
```diff
@@ -212,8 +212,11 @@ static auto BuildFunctionDecl(Context& context,
       parent_scope_inst) {
     if (auto class_decl = parent_scope_inst->TryAs<SemIR::ClassDecl>()) {
       auto& class_info = context.classes().Get(class_decl->class_id);
-      CARBON_CHECK(virtual_modifier != SemIR::Function::VirtualModifier::Impl ||
-                   class_info.is_dynamic);
```
Looks like this used to CHECK-fail on things like
```
class C {
  impl fn F();
}
```
Would be nice to add a test for that.

dwblaikie replied:
Oh, sure - added in df58bcda2bf15c117b9ed76bf69a0323b9ac450c

zygoloid commented on toolchain/check/handle_function.cpp:
```diff
@@ -212,8 +212,11 @@ static auto BuildFunctionDecl(Context& context,
+      if (virtual_modifier == SemIR::Function::VirtualModifier::Impl &&
+          !class_info.base_id.is_valid()) {
+        CARBON_DIAGNOSTIC(ImplWithoutBase, Error, "impl without base class");
+        context.emitter().Build(node_id, ImplWithoutBase).Emit();
```
I suppose we'll eventually want to check for a *dynamic* base class here, and then check that the base class has a matching virtual function. This seems like a good step towards that.

dwblaikie replied:
Yep. Given your comment I thought about whether I could/should add the check that the base was dynamic in this patch - but thinking about it, I'm not sure we'd want to describe that to a user (it'd amount to "there are no virtual functions in the base class", I think? - which is not a distinctly interesting property for the user, probably, and would fall out of/probably use the same diagnostic as "we couldn't find a virtual function with this name in any base" (that there were /no/ virtual functions at all doesn't seem like an interesting distinction))

zygoloid reviewed (commented):
Thanks, this seems reasonable.

I assume we don't want to allow `impl fn` to *precede* the `extend base` declaration that introduces the bae class? (I'm not sure whether we've considered that in the language rules but it seems reasonable to require the base is introduced before its virtual functions are overridden.)

dwblaikie committed:
Add simple impl-without-base test

dwblaikie commented:
> I assume we don't want to allow impl fn to precede the extend base declaration that introduces the bae class? (I'm not sure whether we've considered that in the language rules but it seems reasonable to require the base is introduced before its virtual functions are overridden.)

Oh, I don't think we check for that. We had decided that base has to come before fields, but hadn't discussed base before functions. I think that makes just as much sense - I'll work on that diagnostic too, in a separate patch. (maybe base before all functions, to keep things simple? (we might want to check for shadowing, etc) Or did you want only base before virtual functions?)

dwblaikie commented:
Added a test and removed CHECK for the case where the impl function's class doesn't have a dynamic base - so we don't CHECK-fail, but left it as a todo in the code and the test to diagnose the problem.

zygoloid commented:
> (maybe base before all functions, to keep things simple? (we might want to check for shadowing, etc) Or did you want only base before virtual functions?)

I'm inclined to say base before virtual functions only for now. We had some desire to allow at least nested classes to precede the base declaration, so you can do things like:
```carbon
class MyContainer {
  class MyIterator { ... }
  extend base: ContainerBase(MyIterator);
  // ...
}
```
and I suppose it might also be useful to pass a member function as a generic argument to the base class.

dwblaikie committed:
Fixup tests a bit

zygoloid reviewed (approved):
Thanks! Tests still look wedged; I'm not sure how to unwedge them but maybe trying to merge will help.

dwblaikie committed:
Merge branch 'trunk' into disallow_impl_without_base

dwblaikie committed:
Regen test
